src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,877 @@
+          num_filled++;
+        }
+      }
+    }
+  }
+
+  return num_filled;
+}
+
+void VoxelGrid::fillVoxelSetFromBoolGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                         const std::string& boolGridName,
+                                         bool (*func)(bool)) {
+  assert(grids_.find(boolGridName) != grids_.end());
+  assert(grids_[boolGridName].first == "bool");
+  auto boolGrid = getGrid<bool>(boolGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (func(boolGrid[i][j][k])) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromIntGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                        const std::string& intGridName,
+                                        bool (*func)(int)) {
+  assert(grids_.find(intGridName) != grids_.end());
+  assert(grids_[intGridName].first == "int");
+  auto intGrid = getGrid<int>(intGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (func(intGrid[i][j][k])) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromFloatGrid(std::vector<Mn::Vector3i>& voxelSet,
+                                          const std::string& floatGridName,
+                                          bool (*func)(float)) {
+  assert(grids_.find(floatGridName) != grids_.end());
+  assert(grids_[floatGridName].first == "float");
+  auto floatGrid = getGrid<float>(floatGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (func(floatGrid[i][j][k])) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
+    }
+  }
+}
+
+void VoxelGrid::fillVoxelSetFromVector3Grid(std::vector<Mn::Vector3i>& voxelSet,
+                                            const std::string& vector3GridName,
+                                            bool (*func)(Mn::Vector3)) {
+  assert(grids_.find(vector3GridName) != grids_.end());
+  assert(grids_[vector3GridName].first == "vector3");
+  auto vecGrid = getGrid<Mn::Vector3>(vector3GridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (func(vecGrid[i][j][k])) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
+    }
+  }
+}
+
+// 6D SDF - labels each cell as interior (-inf), exterior (+inf), or boundary
+void VoxelGrid::generateInteriorExteriorVoxelGrid() {
+  // create 6 bool grids
```

Comment:
To me it looks like one int grid could be more efficient, and then use the bits of that int as +X, -X, ... Or, for more type safety and easier use, add a possibility to have a `Math::BoolVector<6>` grid and use that :)

Seeing [Aaron's comment below](https://github.com/facebookresearch/habitat-sim/pull/1140/files#r593914384), it may also help with the code duplication. 

A bit more advanced solution could be massaging the view via `transpose()` / `flip()` so you then always go through the first two dimensions and the third is used for raycasts.